src/nanotron/optim/base.py

```diff
@@ -46,3 +67,112 @@ def inherit_from(self, cls) -> bool:
 
 
 Optimizer = TypeVar("Optimizer", BaseOptimizer, torch.optim.Optimizer)
+
+
+# Modified from torch.optim.Optimizer._process_value_according_to_param_policy
+@staticmethod
+def _process_value_according_to_param_policy(
+    param: torch.Tensor,
+    value: torch.Tensor,
+    param_id: int,
+    param_groups: List[Dict[Any, Any]],
+    map_location: Optional[Union[str, torch.device]],
+    key: Hashable = None,
+) -> torch.Tensor:
+    # If map_location is specified, use it instead of param.device
+    target_device = map_location if map_location is not None else param.device
+
+    fused = False
+    capturable = False
+    assert param_groups is not None
+    for pg in param_groups:
+        if param_id in pg["params"]:
+            fused = pg["fused"] if "fused" in pg else False
+            capturable = pg["capturable"] if "capturable" in pg else False
+            break
+
+    if key == "step":
+        if capturable or fused:
+            return value.to(dtype=torch.float32, device=target_device)
+        else:
+            return value
+    else:
+        if param.is_floating_point():
+            return value.to(dtype=param.dtype, device=target_device)
+        else:
+            return value.to(device=target_device)
+
+
+# Modified from torch.optim.Optimizer.load_state_dict
+@torch._disable_dynamo
+def custom_load_state_dict(self, state_dict: StateDict, map_location: Union[str, torch.device]) -> None:
```

Comment:
It would be nice to avoid duplicating code from PyTorch, in case they introduce internal changes in future versions. Could we create a context manager that automatically moves tensors within a given function to a target device?

```python
@contextmanager
def ensure_cpu():
    original_device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    try:
        torch.set_default_device('cpu')  # This currently requires manual tracking
        yield
    finally:
        torch.set_default_device(original_device)

# Usage
with ensure_cpu():
    tensor = torch.tensor([1.0, 2.0, 3.0])
    print(tensor.device)  # Outputs: cpu

```

Reply:
As discussed, `torch.set_default_device` won't help us create the optimizer states in gpu (optim states are created specifically on the model's device)
I tried a bunch of options before this and none seemed to work. This ensures that we create optim states directly on the device we want